File: golang/15.3Sum.go
const NUM = 3

var pool map[int]int

func threeSum(nums []int) [][]int {
	var answers [][]int
	pool = make(map[int]int)

	sort.Ints(nums)
	answer := make([]int, NUM)
	for _, num := range nums {
		t, exists := pool[num]
		if exists {
			pool[num] = t + 1
		} else {
			pool[num] = 1
		}
	}

	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		pool[nums[i]] = pool[nums[i]] - 1
		for j := i + 1; j < len(nums)-1; j++ {
			if j > i+1 && nums[j] == nums[j-1] {
				continue
			}
			pool[nums[j]] = pool[nums[j]] - 1
			ret := 0 - nums[i] - nums[j]
			if ret >= nums[j] && pool[ret] > 0 {
				answer[0] = nums[i]
				answer[1] = nums[j]
				answer[2] = ret

				result := make([]int, NUM)
				copy(result, answer)
				answers = append(answers, result)
			}
			pool[nums[j]] = pool[nums[j]] + 1
		}
		pool[nums[i]] = pool[nums[i]] + 1
	}

	return answers
}

File: java/15.3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> answers = new ArrayList<>();
        Arrays.sort(nums);

        // build a map
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            map.put(nums[i], map.get(nums[i]) - 1);
            for (int j = i + 1; j < nums.length - 1; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) {
                    continue;
                }
                map.put(nums[j], map.get(nums[j]) - 1);
                int ret = 0 - nums[i] - nums[j];
                if (ret >= nums[j] && map.getOrDefault(ret, 0) > 0) {
                    answers.add(Arrays.asList(nums[i], nums[j], ret));
                }
                map.put(nums[j], map.get(nums[j]) + 1);
            }
            map.put(nums[i], map.get(nums[i]) + 1);
        }

        return answers;
    }
}
